README.md
# JQuery - Simple Validator for Bootstrap

A JQuery plugin that make it easy to validate forms made with Bootstrap.

## Requirements

- JQuery
- Bootstrap 3

## Usage

Simple Validator works based on HTML elements attributes. Using these attributes, you can define, for instance, the biggest number possible or the maximum length for a text in a field. You can also customize the rules for each kind of element, if you need to implement a more complex logic than the supplied one.

Simple validator only checks elements that have either `required="required"` or `data-validate="1"` attributes. The difference is that `data-validate="1"` will allow the field to be empty.

### Basic Attributes

For `input[type="text"]` or `input[type="password"]` or `textarea`:

- `minlength`: define the minimum characters for the input.
- `maxlength`: define the maximum characters for the input.

For `input[type="number"]`:

- `minlength`: define the minimum digits for the input.
- `maxlength`: define the maximum digits for the input.
- `min`: define the floor for the numeric value.
- `max`: define the ceil for the numeric value.

For `input[type="file"]`:

- `data-types`: define the allowed extensions for the file.*

*values must be comma separated, e.g.:

`<input type="file" data-types="jpg,png,gif"/>`

### Example 1 (Default behavior):

When Simple Validator runs, it looks into each form element and evaluate its validation rules. If any of the rules is not being followed, Simple Validator will add a `has-error` class to the container element. Else, Simple Validator will add a `has-success` class to the  container element. You can change this behavior as you prefer.

The following example form is set for the following rules:

- Name must contain between 3 and 32 characters. It's required.
- Email should have a valid email format. It's required.
- Site should have a valid URL format. It's not required.
- Age must be a number between 0 and 120. It's not required.

**HTML:**

```html
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
<script type="text/javascript" src="js/jquery.simplevalidatorbs.min.js"></script>
 <form class="form-horizontal" novalidate="novalidate">
    <div class="form-group">
        <label for="name" class="control-label col-xs-4"><strong>Name:</strong></label>
        <div class="col-xs-8">
            <input class="form-control" id="name" type="text" minlength="3" maxlenght="32" required="required"/>
            <span class="help-block" style="display:none">Invalid name.</span>
        </div>
    </div>
    <div class="form-group">
        <label for="email" class="control-label col-xs-4"><strong>Email:</strong></label>
        <div class="col-xs-8">
            <input class="form-control" id="email" type="email" required="required"/>
            <span class="help-block" style="display:none">Invalid email.</span>
        </div>
    </div>
    <div class="form-group">
        <label for="site" class="control-label col-xs-4"><strong>Site:</strong></label>
        <div class="col-xs-8">
            <input class="form-control" id="site" type="url" data-validate="1"/>
            <span class="help-block" style="display:none">Invalid site.</span>
        </div>
    </div>
    <div class="form-group">
        <label for="age" class="control-label col-xs-4"><strong>Age:</strong></label>
        <div class="col-xs-8">
            <input class="form-control" id="age" type="number" data-validate="1" min="0" max="120"/>
            <span class="help-block" style="display:none">Invalid age.</span>
        </div>
    </div>
    <div class="form-group">
        <div class="col-xs-6">
            <button class="btn btn-block btn-default" type="button" onClick="resetForm()">Reset</button>
        </div>
        <div class="col-xs-6">
            <button class="btn btn-block btn-success" type="submit">Submit</button>
        </div>
    </div>
</form>
```

**JavaScript:**

```javascript
$(function(){
    $('form').submit(function(){;
        var valid = $(this).validateInput();
        return valid;
    });
});
```

### Example 2 (Custom behavior):

Simple Validator evaluate input types (input, textarea, select, etc) and its type attributes (text, number, email, etc) in order to decide what to do. You can pass specific functions for each of these cases. To do this, you have to pass an object ('options') containing one or more of the following arguments: textValidation, numberValidation, emailValidation, passwordValidation, urlValidation, selectValidation, fileValidation, telValidation. Each of these must be a function that returns a boolean regarding the success or failure of the validation proccess.

If you want to change the default behaviour (adding `has-success` or `has-error` class to the field container), you can pass an argument `afterTesting`, with a callback function.

**HTML:**

```html
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
<script type="text/javascript" src="js/jquery.simplevalidatorbs.min.js"></script>
 <form novalidate="novalidate">
    <h3>Demo:</h3>
    <p>This is an example of a form with Simple Validator and custom options. (Try age = 3)</p>
    <div>
        <label for="name"><strong>Name:</strong></label>
        <input id="name" type="text" minlength="3" maxlenght="32" required="required"/>
        <span class="help-block" style="display:none">Invalid name.</span>
    </div>
    <div>
        <label for="email"><strong>Email:</strong></label>
        <input id="email" type="email" required="required"/>
        <span class="help-block" style="display:none">Invalid email.</span>
    </div>
    <div>
        <label for="site"><strong>Site:</strong></label>
        <input id="site" type="url" data-validate="1"/>
        <span class="help-block" style="display:none">Invalid site.</span>
    </div>
    <div>
        <label for="age"><strong>Age:</strong></label>
        <input id="age" type="number" data-validate="1"/>
        <span class="help-block" style="display:none">Invalid age.</span>
    </div>
    <button type="submit">Submit</button>
    <button onClick="reset()">Reset</button>
</form>
```

**JavaScript:**

```javascript
$(function(){
    $('form').submit(function(){
        var valid = $(this).validateInput({
            //define a custom logic for input[type="number"] fields
            'numberValidation' : function(field){
                if (parseInt($(field).val()) == 3) return false;
                return true;
            }
        });
        return valid;
    });
    resetForm = function() {
        //Check Helper Methods for resetInput()
        $('form').resetInput({
            skipSelector: '#name', 
            eachField: function($field){ console.log($field.get(0)) }
        });
    };
});
```

### Options

`afterTesting`
- **Type:** Function(Element field, Boolean valid)
- A function that handles each element after the validation, based on valid.

`textValidation`
- **Type:** Function(Element field) => Boolean
- A function that evaluates input[type="text"] elements on the form and returns true, if valid, or false, if not.

`numberValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="number"] elements on the form and returns true, if valid, or false, if not.

`emailValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="email"] elements on the form and returns true, if valid, or false, if not.

`passwordValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="password"] elements on the form and returns true, if valid, or false, if not.

`urlValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="url"] elements on the form and returns true, if valid, or false, if not.

`selectValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates select elements on the form and returns true, if valid, or false, if not.

`fileValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="file"] elements on the form and returns true, if valid, or false, if not.

`telValidation`
- **Type:** Function(Element field) => Boolean
A function that evaluates input[type="tel"] elements on the form and returns true, if valid, or false, if not.

### Helper Methods

`.resetInput({skipSelector: '', eachFieald: function($field){}})`
-  Reset all inputs in the selected form, except for `skipSelector` ones. If `eachField` is passed in the options, it will run for each field, after the values are empty (JQuery element is passed to callback).

`.hasAttr(String attrName)`
-  Returns `true` or `false` depending on the existence of an attribute on the selected element.

## About

### MIT License

*Permission is hereby granted, free of charge, to any person obtaining a copy * 
of this software and associated documentation files (the "Software"), to    
deal in the Software without restriction, including without limitation the  
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or 
sell copies of the Software, and to permit persons to whom the Software is  
furnished to do so, subject to the following conditions:*                    
                                                                            
*The above copyright notice and this permission notice shall be included in  
all copies or substantial portions of the Software.*                         
                                                                            
*THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR  
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,    
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER      
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING     
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS * 
IN THE SOFTWARE.*

### Author

Ramon Kayo - <contato@ramonkayo.com> - <http://twitter.com/ramonztro>


changelog.md
# Change Log

##[v1.0.0] - 2016-09-24

src/jquery.simplevalidatorbs.js
/*
+---------------------------------------------------------------------------+
| Simple Validator (jquery.simplevalidatorbs.min.js)                          |
+---------------------------------------------------------------------------+
| Author        : Ramon Kayo                                                |
| Email         : contato@ramonkayo.com                                     |
| License       : Distributed under the MIT License                         |
+---------------------------------------------------------------------------+
| "Simplicity is the ultimate sophistication." - Leonardo Da Vinci          |
+---------------------------------------------------------------------------+
| Last modified : 2015-09-24                                                |
+---------------------------------------------------------------------------+
*/
(function ($){
	$.fn.hasAttr = function(attrName) {
		var attr = $(this).attr(attrName);
		return (typeof attr !== 'undefined' && attr !== false);
	};
	$.fn.resetInput = function(options) {
		var settings = {
			skipSelector: '',
			eachField: function(){}
		};

		if (options == null || typeof options != 'object') {
			options = {}
		}
		jQuery.extend(settings, options);

		$(this).find('.has-warning, .has-error, .has-success').removeClass('has-warning has-error has-success');
		$(this).find('.help-block').hide();

		$(this).find('input, select, textarea').each(function(){
			if ($(this).is(settings.skipSelector)) return true;
			
			$(this)
				.val('')
				.prop('selected', false)
				.prop('checked', false);

			settings.eachField($(this));
		});
		return $(this);
	};
	$.fn.validateInput = function(options) {
		var settings = {
			afterTesting : function(field, valid) {
				var $container = $(field).closest('.form-group, .checkbox, .radio');
				var className = valid ? 'has-success' : 'has-error';
				$container
					.removeClass('has-warning has-error has-success')
					.addClass(className);

				if (valid) $container.find('.help-block').hide();
				else $container.find('.help-block').show();
			},
			textValidation : function(field) {
				var value = field.val();
				var length = value.length;
	    		if (!(/\S/.test(value))) return false;
	    		if (field.hasAttr('minlength') && length < field.attr('minlength')) return false; 
	    		if (field.hasAttr('maxlength') && length > field.attr('maxlength')) return false;
	    		return true;
			},
			numberValidation : function(field) {
				console.log('?');
				var value = field.val();
				var length = value.length;
				if (field.hasAttr('minlength') && length < field.attr('minlength')) return false; 
				if (field.hasAttr('maxlength') && length > field.attr('maxlength')) return false;
				value = parseFloat(field.val());
	    		if (isNaN(value))  return false;
	    		if (field.hasAttr('min') && value < field.attr('min')) return false; 
	    		if (field.hasAttr('max') && value > field.attr('max')) return false;
	    		return true;
			},
			emailValidation : function(field) {
				var value = field.val();
				var regex = /^([\w-]+(?:\.[\w-]+)*)@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$/i;
	    		return regex.test(value);
			},
			passwordValidation : function(field) {
				var value = field.val();
				var length = value.length;
	    		if (!(/\S/.test(value))) return false;
	    		if (field.hasAttr('minlength') && length < field.attr('minlength')) return false;
	    		if (field.hasAttr('maxlength') && length > field.attr('maxlength')) return false;
				return true;
			},
			urlValidation : function(field) {
				var value = field.val();
				var regex = /[-a-zA-Z0-9@:%_\+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
			    return regex.test(value);
			},
			selectValidation : function(field) {
				var value = field.val();
				return ($.trim(value) != '');
			},
			fileValidation : function(field) {
				var value = field.val();
	    		var extension = value.split('.').pop().toLowerCase();
	    		var extensions = field.hasAttr('data-types') ? field.data('types').split(',') : [];
	    		if ($.inArray(extension, extensions) == -1) return false;
				return true;
			},
			telValidation : function(field) {
				var value = field.val();
				var regex = /^([0-9()+ -])+$/;
	    		return regex.test(value);
			}
		};
		
		if (options == null || typeof options != 'object') {
			options = {}
		}
		jQuery.extend(settings, options);
		
		var validForm = true;
	    $(this).find('input, select, textarea').each(function() {
	    	if (
	    		!$(this).hasAttr('required') && 
	    		(!$(this).hasAttr('data-validate') || $(this).attr('data-validate') != '1')
	    	) return true;
	    	
	    	var value = $(this).val();
	    	
	    	if (
	    		!$(this).hasAttr('required') && 
	    		(value == null || value == undefined || $.trim(value) == '')
	    	) {
	    		valid = true;
	    	} else {
	    		if ($(this).hasAttr('type')) {
	    			var type = $(this).attr('type').toLowerCase();
	    		} else {
	    			var type = $(this)[0].tagName.toLowerCase();
	    		}
	    		var valid = true;
	    		
	    		if ($(this).data('not-empty') == 1 && $.trim(value) == '') {
	    			valid = false;
	    		} else if (type == 'text' || type == 'textarea') {
	    			valid = settings.textValidation($(this));
	    		} else if (type == 'number') {
	    			valid = settings.numberValidation($(this));
	    		} else if (type == 'email') {
	    			valid = settings.emailValidation($(this));
	    		} else if (type == 'password') {
	    			valid = settings.passwordValidation($(this));
	    		} else if (type == 'url') {
	    			valid = settings.urlValidation($(this));
	    		} else if (type == 'select') {
	    			valid = settings.selectValidation($(this));
	    		} else if (type == 'file') {
	    			valid = settings.fileValidation($(this));
	    		} else if (type == 'tel') {
	    			valid = settings.telValidation($(this));
	    		}
	    	}
	    		
	    	settings.afterTesting(this, valid);
	    	if (!valid) validForm = false;
	    });
	    return validForm;
	};
}(jQuery));